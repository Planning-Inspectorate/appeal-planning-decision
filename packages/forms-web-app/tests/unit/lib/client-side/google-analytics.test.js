/**
 * @jest-environment jsdom
 */
/* eslint-env browser */

const { initialiseGoogleAnalytics } = require('../../../../src/lib/client-side/google-analytics');

describe('lib/client-side/google-analytics', () => {
  const FIXED_SYSTEM_TIME = '2020-11-18T00:00:00Z';
  const fakeGaId = 'some-test-value';
  const gaIdElement = document.createElement('p');
  gaIdElement.id = 'gaId';
  gaIdElement.textContent = fakeGaId;

  const setupFakeDom = () => {
    document.body.append(gaIdElement);
    const script = document.createElement('script');
    document.body.appendChild(script);
  };

  beforeEach(() => {
    jest.resetAllMocks();

    setupFakeDom();

    // https://github.com/facebook/jest/issues/2234#issuecomment-730037781
    jest.useFakeTimers('modern');
    jest.setSystemTime(Date.parse(FIXED_SYSTEM_TIME));

    window.dataLayer = undefined;
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('initialiseGoogleAnalytics when the Google Tag Manager feature flag is false', () => {
    expect(window.dataLayer).toBe(undefined);

    initialiseGoogleAnalytics(document);

    expect(document.body).toMatchSnapshot();

    expect(window.dataLayer).toHaveLength(2);
    // https://github.com/facebook/jest/issues/8475#issuecomment-495729482
    expect([...window.dataLayer[0]]).toMatchObject(['js', new Date()]);
    expect([...window.dataLayer[1]]).toMatchObject(['config', fakeGaId]);
  });

  test('initialiseGoogleAnalytics when the Google Tag Manager feature flag is true', () => {
    document.body.remove(gaIdElement);

    expect(window.dataLayer).toBe(undefined);

    initialiseGoogleAnalytics(document);

    expect(document.body).toMatchSnapshot();

    expect(window.dataLayer).toBe(undefined);
  });
});
