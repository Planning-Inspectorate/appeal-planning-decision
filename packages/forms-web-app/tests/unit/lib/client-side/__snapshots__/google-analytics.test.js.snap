// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/client-side/google-analytics initialiseGoogleAnalytics 1`] = `
<body>
  <p
    id="gaId"
  >
    some-test-value
  </p>
  <script
    src="https://www.googletagmanager.com/gtag/js?id=some-test-value"
    type="text/javascript"
  />
  <script />
</body>
`;
